src/rust/engine/ui/src/console_ui.rs

```diff
@@ -45,6 +46,8 @@ pub struct ConsoleUI {
   local_parallelism: usize,
   // While the UI is running, there will be an Instance present.
   instance: Option<Instance>,
+  teardown_in_progress: bool,
+  teardown_mpsc: (mpsc::Sender<()>, mpsc::Receiver<()>),
```

Comment:
I'm not sure if its worth it / cleaner, but this could be an `Option<(mpsc::Sender<()>, mpsc::Receiver<()>)>` and then you could kill the extra `teardown_in_progress` state.

Reply:
Alternatively, isn't this also modeled by a single `Arc<Mutex<()>>`?

Reply:
I think the `Option<(mpsc::Sender<()>, mpsc::Receiver<()>)>` model would work, but wouldn't really be any cleaner. I like the clarity of a name like `teardown_in_progress`, I think seeing that will make it easier for a newcomer to this code to figure out what's going on here. I'm not sure how I could accomplish something similar with a single `Arc<Mutex<()>>`.